### DISCUSSION TITLE
compiling JS in v5 via gulp

### DISCUSSION THREAD

**Discussion by columbian-chris** [twbs/bootstrap#34342] [Q&A] [answered]
I'm trying to compile a custom JS file with only the components I'm using, but I can't seem to get the load order right or I'm missing something else.  I am trying to compile using this order:

```
node_modules/bootstrap/js/dist/dom/selector-engine.js
node_modules/bootstrap/js/dist/util/index.js
node_modules/bootstrap/js/dist/dom/data.js
node_modules/bootstrap/js/dist/dom/event-handler.js
node_modules/bootstrap/js/dist/base-component.js
node_modules/bootstrap/js/dist/alert.js
node_modules/bootstrap/js/dist/button.js
node_modules/bootstrap/js/dist/dom/manipulator.js
node_modules/bootstrap/js/dist/carousel.js
node_modules/bootstrap/js/dist/collapse.js
node_modules/bootstrap/js/dist/dropdown.js
node_modules/bootstrap/js/dist/util/scrollBar.js
node_modules/bootstrap/js/dist/util/backdrop.js
node_modules/bootstrap/js/dist/modal.js
node_modules/bootstrap/js/dist/offcanvas.js
node_modules/bootstrap/js/dist/util/sanitizer.js
node_modules/bootstrap/js/dist/tooltip.js
node_modules/bootstrap/js/dist/popover.js
node_modules/bootstrap/js/dist/scrollspy.js
node_modules/bootstrap/js/dist/tab.js
node_modules/bootstrap/js/dist/toast.js
```

I got this order from the order they appear in `node_modules/bootstrap/dist/js/bootstrap.js`.  When trying to compile, gulp throws me this error:

```
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: File not found with singular glob: .../node_modules/bootstrap/js/dist/util/index.js (if this was purposeful, use `allowEmpty` option)
    at Glob.<anonymous> (.../node_modules/glob-stream/readable.js:84:17)
    at Object.onceWrapper (events.js:286:20)
    at Glob.emit (events.js:198:13)
    at Glob.EventEmitter.emit (domain.js:448:20)
    at Glob._finish (.../node_modules/glob/glob.js:195:8)
    at done (.../node_modules/glob/glob.js:180:14)
    at Glob._processSimple2 (.../node_modules/glob/glob.js:686:12)
    at .../node_modules/glob/glob.js:674:10
    at Glob._stat2 (.../node_modules/glob/glob.js:770:12)
    at lstatcb_ (.../node_modules/glob/glob.js:762:12)
Emitted 'error' event at:
    at DestroyableTransform.EventEmitter.emit (domain.js:461:12)
    at Pumpify.emit (events.js:198:13)
    at Pumpify.EventEmitter.emit (domain.js:448:20)
    at Pumpify.Duplexify._destroy (.../node_modules/duplexify/index.js:191:15)
    at .../node_modules/duplexify/index.js:182:10
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

In Bootstrap v4, this was the method I was using to accomplish this, but I can't seem to get it right in v5.0.1.

**Comment by 3zrv** [comment]
consider adding `allowEmpty: true` option in gulp.src
Below is a sample code,
```
gulp.src('app/scripts.js', { allowEmpty: true })
.pipe(...);
```
I think this might help you

**Comment by tflight** [comment, marked as answer]
Hi @columbian-chris,

Have a look in the error details:
```Error: File not found with singular glob: .../node_modules/bootstrap/js/dist/util/index.js ```

This means that the file it was trying to include doesn't exist. Indeed the directory `/js/dist/util` doesn't exist in the repository. I suspect the file you are looking for is at `node_modules/bootstrap/js/src/util/index.js`, if you really need it.

**Reply by columbian-chris** [reply]
Oh snap, the `util` dir doesn't actually exist for the `dist` directory.  I took out those files and now I can compile fine.  Thanks.

**Comment by AdamReznicek** [comment]
Hi,
i have to compile only Collapse component via Gulp4. What files and in what order?

`'node_modules/bootstrap/js/dist/dom/selector-engine.js',
'node_modules/bootstrap/js/dist/util/index.js',
      'node_modules/bootstrap/js/dist/dom/data.js',
      'node_modules/bootstrap/js/dist/dom/event-handler.js',
      'node_modules/bootstrap/js/dist/base-component.js',
      'node_modules/bootstrap/js/dist/dom/manipulator.js',
      'node_modules/bootstrap/js/dist/collapse.js',
` 

When i compile, I get the javascript error - Class extends value undefined is not a constructor or null.
Thanks

**Reply by tflight** [reply]
@AdamReznicek try:

```
'node_modules/bootstrap/js/dist/util/index.js',
'node_modules/bootstrap/js/dist/dom/event-handler.js',
'node_modules/bootstrap/js/dist/dom/data.js',
'node_modules/bootstrap/js/dist/dom/manipulator.js',
'node_modules/bootstrap/js/dist/util/config.js',
'node_modules/bootstrap/js/dist/base-component.js',
//'node_modules/bootstrap/js/dist/util/component-functions.js',
'node_modules/bootstrap/js/dist/dom/selector-engine.js',
'node_modules/bootstrap/js/dist/collapse.js',
```

I think `component-functions.js` is often needed, but not in this case. The files above allow the collapse examples to work for me, although there may be another combination that is smaller
